metagraph/core/compiler.py

```diff
@@ -0,0 +1,131 @@
+
+from metagraph.core.plugin import ConcreteAlgorithm, Compiler, CompileError
+from metagraph.core.dask.placeholder import DelayedAlgo
+
+
+@dataclass
+class DaskSubgraph:
+    """A subgraph of a larger Dask task graph.
+
+    Currently subgraph must have 0 or more inputs and only 1 output
+    """
+
+    tasks: dict
+    input_keys: List[Hashable]
+    output_key: Hashable
+
+
+def extract_compilable_subgraphs(
+    dsk: Dict, compiler: str, include_singletons=True
+) -> List[DaskSubgraph]:
+    """Find compilable subgraphs in this Dask task graph.
+
+    Currently only works with one compiler at a time, and only will return
+    linear chains of compilable tasks.  If include_singletons is True,
+    returned chains may be of length 1.  If False, the chain length must be >1.
+    """
+    if include_singletons:
+        chain_threshold = 1
+    else:
+        chain_threshold = 2
+    dependencies, dependents = get_deps(dsk)
+
+    # which nodes are compilable concrete_algorithms?
+    compilable_keys = set()
+    for key in dsk.keys():
+        task_callable = dsk[key][0]
+
+        if (
+            isinstance(task_callable, DelayedAlgo)
+            and task_callable.algo._compiler == compiler
+        ):
+            compilable_keys.add(key)
+
+    non_compilable_keys = set(dsk.keys()) - compilable_keys
+
+    subgraphs = []
+    # keep trying to build chains until no compilable tasks are left
+    keys_left = compilable_keys.copy()
+    while len(keys_left) > 0:
+        key = keys_left.pop()
+        chain = [key]
+        chain_extended = True
+        while chain_extended:
+            next_compilable_tasks = dependents[chain[-1]] & compilable_keys
+            prior_compilable_tasks = dependencies[chain[0]] & compilable_keys
+
+            # Can we extend the chain?
+            chain_extended = False
+
+            if len(next_compilable_tasks) == 1:
+                candidate = next_compilable_tasks.pop()
+                # ensure this task does not have multiple compileable dependencies
+                if len(dependencies[candidate] & compilable_keys) == 1:
+                    chain.append(candidate)
+                    keys_left.remove(chain[-1])
+                    chain_extended = True
+
+            if len(prior_compilable_tasks) == 1:
+                chain.insert(0, prior_compilable_tasks.pop())
+                keys_left.remove(chain[0])
+                chain_extended = True
+
+        if len(chain) >= chain_threshold:
+            # collect all the inputs to the tasks in this chain
+            inputs = set()
+            chain_keys = set(chain)
+            for key in chain:
+                inputs |= dependencies[key] - chain_keys
+            tasks = {key: dsk[key] for key in chain}
+            subgraphs.append(
+                DaskSubgraph(tasks=tasks, input_keys=list(inputs), output_key=chain[-1])
+            )
+
+    return subgraphs
+
+
+def compile_subgraphs(dsk, keys, compiler: Compiler):
```

Comment:
I'm not seeing where the input parameter `keys` is used here. 

Are we intentionally not using it? If so, perhaps we should rename it to `_keys`.

Reply:
Seeing that this parameter is used in https://github.com/metagraph-dev/metagraph/pull/183/files, I think we can resolve this conversation.